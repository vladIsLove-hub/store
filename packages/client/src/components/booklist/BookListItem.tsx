import React, { FC } from 'react';
import { Link } from 'react-router-dom';

import NoneBookImg from '../../icons/OpenBook.svg';
import Star from '../../icons/Star.svg';

import BookListItemCurrency from './BookListItemCurrency';
import { BookListItemProps } from './BookListTypes';

const BookListItem: FC<BookListItemProps> = ({ book }) => {
  const { id, title, price, discount, authors, image, rating } = book;

  const authorsList = authors
    .map(author => {
      return author.name;
    })
    .join(', ');

  return (
    <li data-testid="book-item" className="book-item">
      {book.image ? (
        <img src={image} alt={title} className="book-item__img" />
      ) : (
        <img className="book-item__no-img" src={NoneBookImg} alt="Sorry, image of book was not loaded :C" />
      )}
      <h5 title={title} className="book-list__title">
        {title}
      </h5>
      <BookListItemCurrency price={price} discount={discount} />
      <span title={authorsList} className="book-item__author">
        {authorsList}
      </span>
      <span>
        <img src={Star} className="book-view__star-icon" />
        {rating}/5
      </span>
      <div className="book-item__btns">
        <button className="book-item__addBasket"> Add to Basket </button>
        <Link to={`/books/${id}`} className="book-item__more">
          More
        </Link>
      </div>
    </li>
  );
};

export default BookListItem;
