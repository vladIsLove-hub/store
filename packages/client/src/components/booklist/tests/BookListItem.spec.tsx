import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

import BookListItem from '../BookListItem';

const bookItem = {
  id: '1',
  title: 'Title',
  category: { id: '1', name: 'Science' },
  publisher: { id: '1', name: 'MFTI' },
  authors: [{ id: '1', name: 'Name' }],
  price: 11,
  ageLimit: 16,
  publicationYear: 1999,
  numberOfPages: 321,
  format: ['122', '321'],
  vendor: 'd12r3f',
  rating: 4,
  image: 'ImageURL',
};

describe('BookListItem', () => {
  it('should render correctly', () => {
    const bookWithDiscount = {
      ...bookItem,
      discount: 0.11,
    };
    const { getByTestId, getByText } = render(
      <BrowserRouter>
        <BookListItem key="1" book={bookWithDiscount} />
      </BrowserRouter>,
    );
    const list = getByTestId('book-item');
    expect(list).toHaveClass('book-item');

    expect(list).toContainElement(getByText('Title'));
    expect(list).toContainElement(getByText('4/5'));
    expect(list).toContainElement(getByText('9.79 $'));
    expect(list).toContainElement(getByText('11 $'));
    expect(list).toContainElement(getByText('-11.0 %'));
  });

  it('sould render currently without discount', () => {
    const bookWithoutDiscount = {
      ...bookItem,
      discount: 0,
    };
    const { getByTestId, getByText } = render(
      <BrowserRouter>
        <BookListItem key="1" book={bookWithoutDiscount} />
      </BrowserRouter>,
    );
    const list = getByTestId('book-item');
    expect(list).toHaveClass('book-item');

    expect(list).toContainElement(getByText('Title'));
    expect(list).toContainElement(getByText('4/5'));
    expect(list).toContainElement(getByText('11 $'));
  });

  it('sould render currently without images', () => {
    const bookWithDiscount = {
      ...bookItem,
      discount: 0.11,
      image: '',
    };
    const { getByTestId, getByText, getByAltText } = render(
      <BrowserRouter>
        <BookListItem key="1" book={bookWithDiscount} />
      </BrowserRouter>,
    );
    const list = getByTestId('book-item');
    expect(list).toHaveClass('book-item');

    expect(list).toContainElement(getByText('Title'));
    expect(list).toContainElement(getByText('4/5'));
    expect(list).toContainElement(getByText('9.79 $'));
    expect(list).toContainElement(getByText('11 $'));
    expect(list).toContainElement(getByText('-11.0 %'));
    expect(list).toContainElement(getByAltText('Sorry, image of book was not loaded :C'));
  });

  it('should change location when user click on <more> button', () => {
    const bookWithDiscount = {
      ...bookItem,
      discount: 0.11,
    };
    render(
      <BrowserRouter>
        <BookListItem key="1" book={bookWithDiscount} />
      </BrowserRouter>,
    );
    const pathname = '/books/1';
    userEvent.click(screen.getByText('More'));
    expect(window.location.pathname).toEqual(pathname);
  });

  it('BookListItem snapshot', () => {
    const bookWithDiscount = {
      ...bookItem,
      discount: 0.11,
    };
    const bookListItem = render(
      <BrowserRouter>
        <BookListItem key="1" book={bookWithDiscount} />
      </BrowserRouter>,
    );

    expect(bookListItem).toMatchSnapshot();
  });
});
