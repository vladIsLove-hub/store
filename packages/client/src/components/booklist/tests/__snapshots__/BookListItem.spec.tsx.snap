// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookListItem BookListItem snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="book-item"
        data-testid="book-item"
      >
        <img
          alt="Title"
          class="book-item__img"
          src="ImageURL"
        />
        <h5
          class="book-list__title"
          title="Title"
        >
          Title
        </h5>
        <div
          class="book-item__prices"
          data-testid="currency"
        >
          <span
            class="book-item__price"
          >
            9.79
             $
          </span>
          <span
            class="book-item__overprice"
          >
            11
             $
          </span>
          <span
            class="book-item__percent"
          >
            -
            11.0
             %
          </span>
        </div>
        <span
          class="book-item__author"
          title="Name"
        >
          Name
        </span>
        <span>
          <img
            class="book-view__star-icon"
            src="Star.svg"
          />
          4
          /5
        </span>
        <div
          class="book-item__btns"
        >
          <button
            class="book-item__addBasket"
          >
             Add to Basket 
          </button>
          <a
            class="book-item__more"
            href="/books/1"
          >
            More
          </a>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="book-item"
      data-testid="book-item"
    >
      <img
        alt="Title"
        class="book-item__img"
        src="ImageURL"
      />
      <h5
        class="book-list__title"
        title="Title"
      >
        Title
      </h5>
      <div
        class="book-item__prices"
        data-testid="currency"
      >
        <span
          class="book-item__price"
        >
          9.79
           $
        </span>
        <span
          class="book-item__overprice"
        >
          11
           $
        </span>
        <span
          class="book-item__percent"
        >
          -
          11.0
           %
        </span>
      </div>
      <span
        class="book-item__author"
        title="Name"
      >
        Name
      </span>
      <span>
        <img
          class="book-view__star-icon"
          src="Star.svg"
        />
        4
        /5
      </span>
      <div
        class="book-item__btns"
      >
        <button
          class="book-item__addBasket"
        >
           Add to Basket 
        </button>
        <a
          class="book-item__more"
          href="/books/1"
        >
          More
        </a>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
